lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -50,25 +95,37 @@ class TCV : public std::unary_function<Expr, VisitAction> {
 
   /// Returns known type of \ e
   /// Should be called after visiting the expression to compute its type
-  Expr knownTypeOf(Expr e) {
-    return m_rw->knownTypeOf(e);
-    ;
-  }
+  Expr knownTypeOf(Expr e) { return m_rw->knownTypeOf(e); }
+
+  Expr getErrorExp() { return m_rw->getErrorExp(); }
 };
 } // namespace
 
 namespace expr {
 class TypeChecker::Impl {
+  TCV m_visitor;
+
 public:
+  Impl(TypeChecker *tc) : m_visitor(tc) {}
+
   Expr typeOf(Expr e) {
-    TCV _visitor;
-    visit(_visitor, e);
-    return _visitor.knownTypeOf(e);
+    Expr v = visit(m_visitor, e);
+
+    return m_visitor.knownTypeOf(v);
   }
+
+  Expr getErrorExp() { return m_visitor.getErrorExp(); }
 };
 
-TypeChecker::TypeChecker() : m_impl(new TypeChecker::Impl()) {}
+TypeChecker::TypeChecker() : m_impl(new TypeChecker::Impl(this)) {}
 TypeChecker::~TypeChecker() { delete m_impl; }
 Expr TypeChecker::typeOf(Expr e) { return m_impl->typeOf(e); }
+Expr TypeChecker::getErrorExp() { // to be called after typeOf() or sortOf()
+  return m_impl->getErrorExp();
+}
+void TypeChecker::reset() { 
```

Comment:
Is this an alright way to go about this? The problem that I had was that there isn't a way to know when the user is done with the expression so stuff either has to be reset by the user (like im doing here) or the user has to use a new instance of TypeChecker for a new expression

Reply:
never mind. I'll just store the topmost expression and reset when post-visiting that expression